### DISCUSSION TITLE
Support for storage task upload resume after app restart

### DISCUSSION THREAD

**Discussion by marknelissen** [invertase/react-native-firebase#4678] [Ideas]
The Android SDK supports upload resuming over app restarts, by making use of a upload session id https://firebase.google.com/docs/storage/android/upload-files#continuing_uploads_across_process_restarts . Are there any plans to add support for this into the storage module?

I know of the philosophy of trying to match the web API, but at the same time the aim of this library is to add features that don't exist or work as good in the web SDK. This is one such feature where it would be nice if this could be added. Since it's an addition, the library would still work as a drop-in replacement for the web, but still have this added feature supported.

**Comment by mikehardy** [comment]
There are no current plans, but if you can propose a PR with a reasonable API we could merge it.
We have some precedent of storing things in prefs I believe if you need persistence across restarts 

An semi-highly engineered example (but importantly has the property that it is self-restrained - in a similar way that any solution proposed here would need to provably have bounded storage even if uploads were repeatedly not resumed and thus orphaned session ids)

https://github.com/invertase/react-native-firebase/pull/4317/files#diff-ba8d02468809ff346243fd392fa986246747ad3a7638e1fb24adf515c5f205afR27-R48

**Reply by marknelissen** [reply]
I will see if I have time to put in a (start of) PR. However it would be incomplete, since I have no experience with native iOS development. The basic idea would be to mimic the Android APK, by adding the session url as a third (optional) parameter to the putFile method, and to include the uploadSessionUri in the task snapshot.

When looking at the iOS SDK documentation, it seems that this feature would only work on Android, since only that SDK supports this. Would it then be an option to simply not provide the uploadSessionUri on iOS, and that way the JS code would know if the upload is resumable.

I'm not sure what the position of this project is regarding features that would only fully function on one OS?

**Reply by mikehardy** [reply]
If at all possible it's best to provide sensible defaults to platforms that don't support it *or* fail very fast with a thrown error that developer would catch immediately in development on unsupported platforms

If it is only supportable on Android, I would be willing to merge an android-specific implementation that you attested worked and had a good API, especially if firebase-android-sdk supported it but firebase-ios-sdk does not. Our philosophy in general is "mirror the firebase-js-sdk where possible, where native specifics make it reasonable we can deviate, with the general idea that we will expose the native firebase-android-sdk / firebase-ios-sdk APIs when possible". So if firebase-ios-sdk doesn't have it, then that's that

We have a few platform-specific APIs, where they exist we just mark them clearly in the index.d.ts file (which serve as our reference API) so developers are aware 


Examples:

```
mike@isabela:~/work/Invertase/react-native-firebase (master) % grep -ri 'ios' packages|grep -i 'only'|grep index.d.ts
packages/remote-config/lib/index.d.ts:     * @android Android only - iOS returns Promise<null> but does not reset anything
packages/admob/lib/index.d.ts:     * iOS only
packages/admob/lib/index.d.ts:     * This option is only applied to iOS devices. On Android, this option has no effect.
packages/auth/lib/index.d.ts:     * Apple auth provider implementation. Currently this is iOS only.
packages/auth/lib/index.d.ts:     * iOS only flag to disable app verification for the purpose of testing phone authentication. For this property to take effect, it needs to be set before rendering a reCAPTCHA app verifier. When this is disabled, a mock reCAPTCHA is rendered instead. This is useful for manual testing during development or for automated integration tests.
packages/auth/lib/index.d.ts:     * @param actionCodeSettings The action code settings. The action code settings which provides Firebase with instructions on how to construct the email link. This includes the sign in completion URL or the deep link for mobile redirects, the mobile apps to use when the sign-in link is opened on an Android or iOS device. Mobile app redirects will only be applicable if the developer configures and accepts the Firebase Dynamic Links terms of condition. The Android package name and iOS bundle ID will be respected only if they are configured in the same Firebase Auth project used.
packages/app/lib/index.d.ts:     * iOS only - The OAuth2 client ID for iOS application used to authenticate Google users, for example
packages/app/lib/index.d.ts:     * iOS only - The Android client ID used in Google AppInvite when an iOS app has its Android version, for
packages/app/lib/index.d.ts:     * iOS only - The URL scheme used to set up Durable Deep Link service.
packages/app/lib/index.d.ts:     * @ios iOS only
packages/app/lib/index.d.ts:     * @android Android only - iOS returns null
packages/app/lib/index.d.ts:     * @android Android only - iOS returns null
packages/app/lib/index.d.ts:     * @android Android only - iOS returns 0
packages/app/lib/index.d.ts:     * @android Android only - iOS returns true
packages/app/lib/index.d.ts:     * @android Android only - iOS returns undefined
packages/app/lib/index.d.ts:     * @android Android only - iOS returns undefined
packages/app/lib/index.d.ts:     * @android Android only - iOS returns undefined
packages/app/lib/index.d.ts:     * @android Android only - iOS returns false
packages/app/lib/index.d.ts:     * @android Android only - iOS always returns { isAvailable: true, status: 0 }
packages/app/lib/index.d.ts:     * @android Android only - iOS always returns { isAvailable: true, status: 0 }
packages/app/lib/index.d.ts:     * @android Android only - iOS returns undefined
packages/app/lib/index.d.ts:     * @android Android only - iOS returns undefined
packages/app/lib/index.d.ts:     * @android Android only - iOS returns undefined
```

Hopefully this makes sense + helps? It is always my pleasure to shepherd community PRs to successful merge, that is the heart of open source work, I'll help any way I can
